### PR TITLE
JIT: fix another jitstress/gcstress issue with profiler leave hook

### PR DESCRIPTION
**Pull request by AndyAyersMS**
Need to also report the hidden return buffer pointer as GC live, in case it
happens to refer to a heap location (method invoked via reflection).

Also did some minor cleanup; we weren't using `returnsGCr` in `emitEndCodeGen`
and we were munging return value liveness unnecessarily when not emitting
a profiler leave hook.

Fixes remaining cases in #1971.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@BruceForstall PTAL
cc @dotnet/jit-contrib 

Further cleanup in this area deferred; tracked via #2104

**Comment by AndyAyersMS** [comment]
Added runtime-coreclr gcstress-extra runs...

File: src/coreclr/src/jit/emit.cpp [hunk 1]
```diff
@@ -4442,17 +4442,33 @@ void emitter::emitComputeCodeSizes()
...
+//    consAddr [OUT] - address of the read only constant buffer (if any)
+//
+// Notes:
+//    Currently, in methods with multiple epilogs, all epilogs must have the same
+//    size. epilogSize is the size of just one of these epilogs, not the cumulative
+//    size of all of the method's epilogs.
+//
+// Returns:
+//    size of the method code, in bytes
+//
```
Comment (CarolEidt) [inline src/coreclr/src/jit/emit.cpp]: Excellent - thanks for the comment!

**Review by CarolEidt — APPROVED** [review]
LGTM

**Comment by AndyAyersMS** [comment]
New control logic in `genReturn` is not quite right, seeing the new assert firing in stress cases.

**Comment by AndyAyersMS** [comment]
Suspect I just need to screen out RETFILT, but waiting to see if a repro comes up on winx64 first.
```C++
    if (compiler->compIsProfilerHookNeeded()) // && (treeNode->OperGet() == GT_RETURN)
    {
        // There should be a single return block while generating profiler ELT callbacks.
        assert(compiler->compCurBB == compiler->genReturnBB);
```

**Comment by AndyAyersMS** [comment]
Pushed a fix; re-added runtime-coreclr gcstress-extra runs...

**Comment by AndyAyersMS** [comment]
Still seeing the new assert fire in a few x86 jitstress/gcstress cases. An example of this is in `Bmi2.X64_r.dll` for `X64:MultiplyNoFlags(long,long,int):long`

`MultiplyNoFlags` is not supported on x86 so we get a comma-wrapped throw out of  the importer:
```
 *  RETURN    long  
 \--*  COMMA     long  
    +--*  CALL help void   HELPER.CORINFO_HELP_THROW_PLATFORM_NOT_SUPPORTED
    \--*  LCL_VAR   long   V03 tmp0         
```
Under jitstress we decide to add profile enter and exit hooks. So`fgAddInternal` produces a canonical return BB which is must keep.

Morph converts the entry BB to BBJ_THROW but does not unwrap the comma. The canoncial return block stays around even though nothing targets it. Rationalize expands out the return.
```
-- BB01 [000..009) (throw)

          IL_OFFSET void   IL offset: 0x0
          CALL help void   HELPER.CORINFO_HELP_THROW_PLATFORM_NOT_SUPPORTED $141
 t12 =    LCL_VAR   long   V03 tmp0         u:1 (last use) $80
       /--*  t12    long   
       *  RETURN    long   $1c0

-- BB02 [???..???) (return)

 t14 =    LCL_VAR   long   V06 tmp3          (last use)
       /--*  t14    long   
       *  RETURN    long  
```
and so now we have a GT_RETURN in a block that is not the canonical return block, hence the assert.

Could fix this in a number of ways, but checking that the block is BBJ_RETURN should work.

Remainder of the ~120 odd failures in runtime-coreclr gcstress-extra are on arm/arm64; a quick gloss over these gives the impression the arm failures are similar to the ones in #1971, the but arm64 cases seem like general jitstress failures.

**Comment by AndyAyersMS** [comment]
Decided to just go back to something closer to the original -- only put the profiler exit hook in the canonical return BB.

**Comment by AndyAyersMS** [comment]
runtime-coreclr gcstress-extra looking better now; 135 failures remain (down from ~600 last weekend), only two failures left on windows (x86).

**Comment by BruceForstall** [comment]
The Linux arm64 failures look like failures I've seen in the "runtime-coreclr jitstress" pipeline, but I haven't been able to reproduce them.

**Comment by BruceForstall** [comment]
Oh, and there are 2 distinct JIT asserts, all in the hwintrinsics tests.

**Comment by AndyAyersMS** [comment]
The remaining x86 failure: looks like `bextr` is not killing gcness of its destination. Will look into a fix.

**Comment by BruceForstall** [comment]
Doh! Once again forgot to set `COMPlus_TieredCompilation=0`...

**Comment by BruceForstall** [comment]
New issues to cover the Linux/arm64 failures:
https://github.com/dotnet/runtime/issues/2181
https://github.com/dotnet/runtime/issues/2182

**Pull request merged**